### ISSUE TITLE
Entities: identity, class, and type

### ISSUE THREAD

**Issue by byorgey** [swarm-game/swarm#455]
The problem
-----------------

There are several open issues struggling with how to deal with entities when we need to be able to distinguish among entities that would otherwise be considered identical.

- #114 wonders about how to manage things like `box`es with things put inside them: once we can start putting things in boxes, it matters a lot *which* box we are referring to.  For example, we want to be able to say things like "take one of my 50 `box` entities, put a bunch of things in it, and now give that specific box to another robot".
- #116 wonders about how to distinguish among different `paper` entities that have had values printed on them.

I'm sure other, similar situations will arise as we continue designing more interesting entities and commands.

In general, we have considered entities to be uniquely identified by their *name*.  However, this causes a lot of problems.
- Entities aren't *actually* uniquely identified or stored by name, but rather by hash value.  I think there are multiple places in the code where we look up an entity by name and just *assume* that we will get only one, even though in theory there could be multiple.
- Changing the name of an entity to make it unique is problematic.  Consider the example of a `box` again.  #114 suggests taking a `box` entity and changing its name to something unique (*e.g.* `box1425`) so that we can now refer to it uniquely.  The problem is that we now no longer know that it is a `box`!  As in, we might want to say something like "give every box in my inventory to `base`", but we can't: the `box1425` is no longer a `box`, it is now a `box1425`.
- Moreover, representing entities as `string` values is error prone and not type-safe: e.g. `give base "treee"` only fails at runtime.

Proposed solution concept
-------------------------------------

The germ of this idea came from @noahyor .  #393 noticed that principles of object-oriented programming were helpful in thinking about the way robots interact; the same thing is true here.  The basic idea is to formally distinguish between the *class* of an entity (represented by a `string` such as `box`) and its *identity* (represented by some kind of internal `int` ID, probably).  For efficiency, we can still have multiple identical entities represented simply by a count (for example, if we have harvested 1257 `tree`s, we do not need to give each one a separate, unique identity! --- we just have 1257 copies of the generic `tree` entity).  But we can also have multiple entities with distinct identities but the same `class`.  This makes the `box` and `paper` examples above work cleanly.  For example, we can refer to a specific `box` by its unique identity, but it is still a `box`, so we can do things like "give all `box`es" even if some of them have distinct identities.

Proposed solution details
----------------------------------

- Make a new opaque `entity` type, similar to what we did with robots in #303 .
    - Internally, an `entity` value can be stored as an `Int`, the entity hash value we already compute.
    - `entity` values must be *immutable*, since changing an attribute of an entity changes its hash value and thus its identity.
    - Some built-in functions will change type as well.  For example, `give : robot -> entity -> cmd ()`, `place : entity -> cmd ()`, and `grab : cmd entity`.
    - `make`, on the other hand, should probably *not* change type; more on this below.
- The `entityMap` (and `entities.yaml`) should now be thought of as storing a *prototype* `Entity` object corresponding to each entity class name.
- There should be a function like `entity : string -> entity` (alternative name: `proto`?) which does a lookup in the entity map, *i.e.* returns the prototype entity associated with a particular class name.
    - Presumably it throws an exception if there is no class with the given name.
    - So *e.g.* if we want to give the base one of our generic trees we can say `give base (entity "tree")`.
        - This is slightly annoying but of course everyone is free to define their own `def give' = \r. \c. give r (entity c) end`.
        - Of course one can also `thing <- grab; give base thing` just like before.
    - But if we have a variable `b : entity` representing, say, a special, specific box with some things put in it, we can also say `give base b`.
- Note that we could still have `make : string -> cmd ()` since it only ever makes sense for `make` to produce a prototype entity of a given class.
- There should also be a function `class : entity -> string`.
- If you have a specific `e : entity` and you want to `make` another one of it, you can of course say `make (class e)`.
- There will be a function like `getUnique : string -> cmd entity`.
    - Note this has to be in the `cmd` monad, since it will
        - Find and *remove* a prototype entity of the given class from your inventory
        - *Add* a version of that entity with a new, unique id to your inventory
        - Return a reference to the new, unique entity
    - This is how we can solve the problems posed in #114 and #116 .
        - For example, `myBox <- getUnique "box"; insert (entity "tree") mybox; give base myBox`.
        - Likewise, `print : a -> cmd entity` will remove a prototype `paper` entity from your inventory and return a new, unique entity of the same class.
- Note that `give base (entity "box")` will only ever give prototype boxes, never boxes returned by `getUnique`, because they have a different `entity` value.  So how can we do something like "give the base all the entities with class `box`"?  A few thoughts:
    - The simplest idea is to have a function like `getEntity : string -> cmd entity` which returns a reference to *any* entity of the given class in your inventory, as opposed to `entity` which specifically only returns a prototype entity.
    -  Likewise there should probably be `has : entity -> cmd bool` vs `hasAny : string -> cmd bool` and `count : entity -> cmd bool` vs `countAny : string -> cmd bool`.
        - Very open to bikeshedding on these and other names. :smile:
    - Hence we could write something like
        ```
        while (hasAny "box") { e <- getEntity "box"; give base e }
        ```

**Comment by xsebek** [comment]
I like the detailed proposal 👍 

Would it be possible to have automatic conversion like Haskell `fromIntegral` that would convert string to entity where necessary?
Not sure if we want that, but it would be less to write.

I like the expandable item idea. Grouping items sounds useful on its own, so I may write an Issue for it. 😁

**Comment by xsebek** [comment]
Thinking about this a bit, do we need the random generation of entity UUIDs when we have hashes?

Or would this overhaul remove the current entity hashing system?

**Comment by xsebek** [comment]
Re showing the numbers, we could show the name and as little of the unique identifier (or hash) as necessary to disambiguate them for current robot.

Similarly to how Git shows longer SHA as the repo history increases.

**Comment by byorgey** [comment]
No, I was talking about unique id's just as a sort of hypothetical idea, which I then explained that we *don't* want.  We definitely want to just use hashes.

> Re showing the numbers, we could show the name and as little of the unique identifier (or hash) as necessary to disambiguate them for current robot.

That's a good idea.

**Comment by xsebek** [comment]
Now that records are part of the game, could we use them for entities?

That way we might avoid adding many custom commands just for entities.

**Comment by byorgey** [comment]
> Now that records are part of the game, could we use them for entities?

Can you elaborate?  I'm not sure I understand what you mean.

**Comment by xsebek** [comment]
@byorgey something like `type entity = [name: String, display: Display,...]` so we could access information about entities directly.

**Comment by byorgey** [comment]
Oh, I see, that's a nice idea. Though to make it ergonomic would require #153 .  There are probably other hidden gotchas, or extra features we would want for records to make it work well, but it's definitely worth exploring.

**Comment by byorgey** [comment]
Random idea I had, in reference to this question:

> Would it be possible to have automatic conversion like Haskell fromIntegral that would convert string to entity where necessary?

Above I answered "no"; however, in thinking about it more I think a compromise might be possible: we could allow text literals to *check* at type `entity`, but without adding any subtyping or qualified polymorphism.  That is, if you ask for the type of `"tree"` it will still say `Text`, and you still could not, say, compose a function that returns `Text` with a function that expects `Entity`.  However, it would help in the common case that you are writing a literal entity name.  For example you could write `give base "tree"`, and since the type of `give base` is inferred to be `Entity -> Cmd Unit`, it would *check* that `"tree"` has type `Entity`, which would succeed.

**Comment by xsebek** [comment]
@byorgey so you are suggesting something like OverloadedStrings?

**Comment by byorgey** [comment]
No, `OverloadedStrings` actually works by wrapping string literals in `fromString :: IsString a => String -> a`, giving them the qualified polymorphic type `IsString a => a`.  GHC then has to do type inference to figure out what the type `a` should be.  I am proposing to *not* do that (which would be very complex).

What I am proposing is much simpler: we always assume text literals have type `Text`, unless the *immediate context* expects the type `Entity` (*i.e.* we apply some `Entity -> ...` function to a text literal) in which case we allow that too.
